auth/opa/rpcauth/rpcauth.go

```diff
@@ -88,6 +94,110 @@ func NewWithPolicy(ctx context.Context, policy string, authzHooks ...RPCAuthzHoo
 	return New(p, authzHooks...), nil
 }
 
+func isMessage(descriptor protoreflect.FieldDescriptor) bool {
+	return descriptor.Kind() == protoreflect.MessageKind || descriptor.Kind() == protoreflect.GroupKind
+}
+
+func isDebugRedactEnabled(fd protoreflect.FieldDescriptor) bool {
+	opts, ok := fd.Options().(*descriptorpb.FieldOptions)
+	if !ok {
+		return false
+	}
+	return opts.GetDebugRedact()
+}
+
+func redactListField(value protoreflect.Value) {
+	for i := 0; i < value.List().Len(); i++ {
+		redactFields(value.List().Get(i).Message())
+	}
+}
+
+func redactMapField(value protoreflect.Value) {
+	value.Map().Range(func(mapKey protoreflect.MapKey, mapValue protoreflect.Value) bool {
+		redactFields(mapValue.Message())
+		return true
+	})
+}
+
+func redactNestedMessage(message protoreflect.Message, descriptor protoreflect.FieldDescriptor, value protoreflect.Value) {
+	switch {
+	case descriptor.IsList() && isMessage(descriptor):
+		redactListField(value)
+	case descriptor.IsMap() && isMessage(descriptor):
+		redactMapField(value)
+	case !descriptor.IsMap() && isMessage(descriptor):
+		redactFields(value.Message())
+	}
+}
+
+func redactSingleField(message protoreflect.Message, descriptor protoreflect.FieldDescriptor) {
+	if descriptor.Kind() == protoreflect.StringKind {
+		if descriptor.Cardinality() != protoreflect.Repeated {
+			message.Set(descriptor, protoreflect.ValueOfString("--REDACTED--"))
+		} else {
+			list := message.Mutable(descriptor).List()
+			for i := 0; i < list.Len(); i++ {
+				list.Set(i, protoreflect.ValueOfString("--REDACTED--"))
+			}
+		}
+	} else {
+		// other than string, clear it
+		message.Clear(descriptor)
+	}
+}
+
+func redactFields(message protoreflect.Message) {
+	message.Range(
+		func(descriptor protoreflect.FieldDescriptor, value protoreflect.Value) bool {
+			if isDebugRedactEnabled(descriptor) {
+				redactSingleField(message, descriptor)
+				return true
+			}
+			redactNestedMessage(message, descriptor, value)
+			return true
+		},
+	)
+}
+
+const mockMessageType = "Mock.MockRequest"
+
+func getRedactedInput(input *RPCAuthInput) (RPCAuthInput, error) {
+	if input == nil {
+		return RPCAuthInput{}, nil
+	}
+	redactedInput := RPCAuthInput{
+		Method:      input.Method,
+		MessageType: input.MessageType,
+		Metadata:    input.Metadata,
+		Peer:        input.Peer,
+		Host:        input.Host,
+		Environment: input.Environment,
+		Extensions:  input.Extensions,
+	}
+	if input.MessageType == mockMessageType || input.MessageType == "" {
```

Comment:
Why do we need the special case for Mock?

If it's for getting around the tests passing in fake proto types, I'd rather fix that by making them pass around real proto types. You can use google.protobuf.Empty for cases where you want to leave input.Message blank.

Reply:
Thanks for catching this. Yeah I changed it to use protobuf.Empty and removed the special casing.